### ISSUE TITLE
MPICH large_type_sendrec hangs with TCP BTL

### ISSUE THREAD

**Issue by artpol84** [open-mpi/ompi#1174]
@bosilca @rhc54 @jsquyres 

I see 100% reproducable hangs of the `large_type_sendrec` test. As far as I can tell, in this test only first and last ranks are actually communicating, all others are just waiting in `MPI_Finalize()`.
This test has an agrument that says the power of 2 for the sen/recv buffer size. I see the following picture:

| argument | time |
| --- | --- |
| 30 | 13 s |
| 31 | 27 s |
| 32 | **hang** |

If I use openib or vader BTL's it works fine for all mentioned buffer sizes.

Backtraces shows the following (I provide only the first and the last ranks):
**rank=0**

``` bash
(gdb) bt
#0  0x00000037feaaccdd in nanosleep () from /lib64/libc.so.6
#1  0x00000037feae1e54 in usleep () from /lib64/libc.so.6
#2  0x00007f4d2700889c in OPAL_PMIX_PMIX1XX_PMIx_Fence (procs=0x0, nprocs=0, info=0x0, ninfo=0) at src/client/pmix_client_fence.c:100
#3  0x00007f4d27238f3c in pmix1_fence (procs=0x0, collect_data=0) at pmix1_client.c:306
#4  0x00007f4d28f1622d in ompi_mpi_finalize () at runtime/ompi_mpi_finalize.c:242
#5  0x00007f4d28f40b2d in PMPI_Finalize () at pfinalize.c:47
#6  0x0000000000401764 in main (argc=2, argv=0x7fff66de9f18) at large_type_sendrec.c:182
```

**rank=16 (last)**

```
(gdb) bt
#0  0x0000003c000cf527 in sched_yield () from /lib64/libc.so.6
#1  0x00007f3adbc60b79 in opal_progress () at runtime/opal_progress.c:199
#2  0x00007f3adc2c7508 in opal_condition_wait (c=0x7f3adc5f3f00, m=0x7f3adc5f3e80) at ../opal/threads/condition.h:76
#3  0x00007f3adc2c75ae in ompi_request_wait_completion (req=0x18d5a00) at ../ompi/request/request.h:383
#4  0x00007f3adc2c75ff in ompi_request_default_wait (req_ptr=0x7fff068dbc68, status=0x7fff068dbc48) at request/req_wait.c:42
#5  0x00007f3adc324830 in PMPI_Wait (request=0x7fff068dbc68, status=0x7fff068dbc48) at pwait.c:71
#6  0x00000000004015db in main (argc=2, argv=0x7fff068dbdb8) at large_type_sendrec.c:151
```

**Comment by bosilca** [comment]
I can reliably reproduce this on OS X. The problem is that the datatype engine identifies the buffer as contiguous, and passes the entire 2^32 bytes buffer to writev. The OS X [writev man page](https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man2/write.2.html)  does highlight that the sum of iov_len values in the iov array should not overflows a 32-bit integer, which explains why we trigger the EINVAL return from writev.

Are you seeing a similar issue on other architectures?

**Comment by artpol84** [comment]
Mine is on RHEL.

четверг, 3 декабря 2015 г. пользователь bosilca написал:

> I can reliably reproduce this on OS X. The problem is that the datatype
> engine identifies the buffer as contiguous, and passes the entire 2^32
> bytes buffer to writev. The OS X writev man page
> <https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man2/write.2.html>
> does highlight that the sum of iov_len values in the iov array should not
> overflows a 32-bit integer, which explains why we trigger the EINVAL return
> from writev.
>
> Are you seeing a similar issue on other architectures?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/open-mpi/ompi/issues/1174#issuecomment-161473643>.
>


-- 
-----
Best regards, Artem Polyakov
(Mobile mail)

**Comment by bosilca** [comment]
What is the man page of writev saying about the cases where EINVAL is returned ?

**Comment by artpol84** [comment]
```
ERRORS
       The  errors  are  as  given for read(2) and write(2).  Furthermore, preadv() and pwritev() can also fail for the same reasons as
       lseek(2).  Additionally, the following error is defined:

       EINVAL The sum of the iov_len values overflows an ssize_t value.  Or, the vector count iovcnt is less than zero or greater  than
              the permitted maximum.
```

**Comment by artpol84** [comment]
`sizeof(ssize_t) = 8` on my system. So 2^32 shouldn't be a problem.

**Comment by bosilca** [comment]
Set a breakpoint in btl_tcp_frag.c line 132 on your last rank. Once it triggers look at the content of the iov and access all elements (all iov_count as well as the first and last element of each iov_base).

**Comment by artpol84** [comment]
@bosilca 
this line: https://github.com/open-mpi/ompi/blob/master/opal/mca/btl/tcp/btl_tcp_frag.c#L132 never triggers. Note, that rank 0 is actually sending the data and rank 1 (in case of 2 procs) is receiver.
When I put a breakpoint on https://github.com/open-mpi/ompi/blob/master/opal/mca/btl/tcp/btl_tcp_frag.c#L117 it indeed has triggered in rank 1, return value was > 0 and iov content was the following:
```
#0  mca_btl_tcp_frag_send (frag=0x7ffff2381b00, sd=18) at btl_tcp_frag.c:117
117             cnt = writev(sd, frag->iov_ptr, frag->iov_cnt);
(gdb) p frag->iov_cnt
$2 = 2
(gdb) p frag->iov_ptr[0]
$3 = {iov_base = 0x7ffff2381bb8, iov_len = 8}
(gdb) p frag->iov_ptr[1]
$4 = {iov_base = 0x7ffff2381c58, iov_len = 56}
```

**Comment by ggouaillardet** [comment]
@artpol84 looks like an overflow ... i am now testing this
```patch
diff --git a/opal/mca/btl/tcp/btl_tcp_frag.c b/opal/mca/btl/tcp/btl_tcp_frag.c
index d9bf9b7..2cdd1b5 100644
--- a/opal/mca/btl/tcp/btl_tcp_frag.c
+++ b/opal/mca/btl/tcp/btl_tcp_frag.c
@@ -109,7 +109,7 @@ size_t mca_btl_tcp_frag_dump(mca_btl_tcp_frag_t* frag, char* msg, char* buf, siz
 
 bool mca_btl_tcp_frag_send(mca_btl_tcp_frag_t* frag, int sd)
 {
-    int cnt=-1;
+    ssize_t cnt=-1;
     size_t i, num_vecs;
 
     /* non-blocking write, but continue if interrupted */
@@ -142,7 +142,7 @@ bool mca_btl_tcp_frag_send(mca_btl_tcp_frag_t* frag, int sd)
     /* if the write didn't complete - update the iovec state */
     num_vecs = frag->iov_cnt;
     for(i=0; i<num_vecs; i++) {
-        if(cnt >= (int)frag->iov_ptr->iov_len) {
+        if(cnt >= frag->iov_ptr->iov_len) {
             cnt -= frag->iov_ptr->iov_len;
             frag->iov_ptr++;
             frag->iov_idx++;
```

**Comment by ggouaillardet** [comment]
@artpol84 @bosilca 
here is a better version but with printf debug
```patch
diff --git a/opal/mca/btl/tcp/btl_tcp_frag.c b/opal/mca/btl/tcp/btl_tcp_frag.c
index d9bf9b7..ea394b2 100644
--- a/opal/mca/btl/tcp/btl_tcp_frag.c
+++ b/opal/mca/btl/tcp/btl_tcp_frag.c
@@ -109,12 +109,14 @@ size_t mca_btl_tcp_frag_dump(mca_btl_tcp_frag_t* frag, char* msg, char* buf, siz
 
 bool mca_btl_tcp_frag_send(mca_btl_tcp_frag_t* frag, int sd)
 {
-    int cnt=-1;
+    ssize_t cnt=-1;
     size_t i, num_vecs;
 
     /* non-blocking write, but continue if interrupted */
     while(cnt < 0) {
+fprintf(stderr,"writev(%ld)...\n", frag->iov_ptr[frag->iov_cnt-1].iov_len);
         cnt = writev(sd, frag->iov_ptr, frag->iov_cnt);
+fprintf(stderr,"writev(%ld) returned %ld\n", frag->iov_ptr[frag->iov_cnt-1].iov_len, cnt);
         if(cnt < 0) {
             switch(opal_socket_errno) {
             case EINTR:
@@ -142,7 +144,7 @@ bool mca_btl_tcp_frag_send(mca_btl_tcp_frag_t* frag, int sd)
     /* if the write didn't complete - update the iovec state */
     num_vecs = frag->iov_cnt;
     for(i=0; i<num_vecs; i++) {
-        if(cnt >= (int)frag->iov_ptr->iov_len) {
+        if((size_t)cnt >= frag->iov_ptr->iov_len) {
             cnt -= frag->iov_ptr->iov_len;
             frag->iov_ptr++;
             frag->iov_idx++;
@@ -159,7 +161,8 @@ bool mca_btl_tcp_frag_send(mca_btl_tcp_frag_t* frag, int sd)
 
 bool mca_btl_tcp_frag_recv(mca_btl_tcp_frag_t* frag, int sd)
 {
-    int cnt, dont_copy_data = 0;
+    int dont_copy_data = 0;
+    ssize_t cnt;
     size_t i, num_vecs;
     mca_btl_base_endpoint_t* btl_endpoint = frag->endpoint;
 
@@ -237,7 +240,7 @@ bool mca_btl_tcp_frag_recv(mca_btl_tcp_frag_t* frag, int sd)
     /* if the read didn't complete - update the iovec state */
     num_vecs = frag->iov_cnt;
     for( i = 0; i < num_vecs; i++ ) {
-        if( cnt < (int)frag->iov_ptr->iov_len ) {
+        if( (size_t)cnt < frag->iov_ptr->iov_len ) {
             frag->iov_ptr->iov_base = (opal_iov_base_ptr_t)
                 (((unsigned char*)frag->iov_ptr->iov_base) + cnt);
             frag->iov_ptr->iov_len -= cnt;
```

the program hangs ...
on the sender size, we kind of enter an infinite loop
writev(...,{iov_base=xxx,iov_len=1<<32},1) that always returns 0 (!)

this is a rhel6 cluster

**Comment by artpol84** [comment]
@bosilca @ggouaillardet 
I'm observing receiver side hang, sender sleeps in MPI_Finlaize();

**Comment by ggouaillardet** [comment]
fwiw, here is an additional patch that works for me
this is a quick and dirty proof of concept no one should use for real ...
if instead of one ```iov``` with ```iov_len``` = 1<<32 i use two ```iov``` with ```iov_len``` = 1<<31 it works (!)

```
diff --git a/opal/mca/btl/tcp/btl_tcp.c b/opal/mca/btl/tcp/btl_tcp.c
index f618a0a..7997988 100644
--- a/opal/mca/btl/tcp/btl_tcp.c
+++ b/opal/mca/btl/tcp/btl_tcp.c
@@ -398,6 +398,12 @@ int mca_btl_tcp_put (mca_btl_base_module_t *btl, struct mca
         frag->iov[i+2].iov_base = (IOVBASE_TYPE*)frag->segments[i].seg_addr.pva
         frag->iov_cnt++;
     }
+    if (frag->iov[frag->iov_cnt-1].iov_len == 0x80000000) {
+        frag->iov[frag->iov_cnt-1].iov_len = 0x40000000;
+        frag->iov_cnt++;
+        frag->iov[frag->iov_cnt-1].iov_len = 0x40000000;
+        frag->iov[frag->iov_cnt-1].iov_base = frag->iov[frag->iov_cnt-2].iov_ba
+    }
     frag->hdr.base.tag = MCA_BTL_TAG_BTL;
     frag->hdr.type = MCA_BTL_TCP_HDR_TYPE_PUT;
     frag->hdr.count = 1;
```

**Comment by ggouaillardet** [comment]
@artpol84 
the sender is hanging in MPI_Finalize because it think it sent the message ... but it did not
(see my second patch that fixes the overflow, so the sender do not incorrectly believe the message was sent)

**Comment by artpol84** [comment]
@bosilca @ggouaillardet 
I've made some more digging and found that since I have 2 TCP interfaces on the node the message was split onto 2 parts 2^31 each. And still I'm seeing hangs!
Also I see that `writev` returns way smaller bytecount (something like 200K) and according to the man page it returns bytes sent. I'll try to find out what happens.

**Comment by artpol84** [comment]
@bosilca @ggouaillardet 
please, ignore prev comment (about writev) - I missed that it is processed later.

**Comment by ggouaillardet** [comment]
i will double check that tomorrow, it seems i got confused with 1<<31 and 1<<32
And i might have two interfaces too ... (eth0 and ib0)

**Comment by ggouaillardet** [comment]
btw, the test works fine if running on one node with ```--mca btl tcp,self```
could we be hitting a limit specific to eth0 and/or ib0 ?

**Comment by artpol84** [comment]
@bosilca @ggouaillardet 
Gilles is right about his first patch. @ggouaillardet I suggest you to create a PR on it.
The reason why he saw the hang was here:
https://github.com/open-mpi/ompi/blob/master/ompi/mca/pml/ob1/pml_ob1_recvreq.c#L205

Second argument of `OPAL_THREAD_ADD_SIZE_T` is **wrong**:
```C
static inline size_t OPAL_THREAD_ADD_SIZE_T(volatile size_t *addr, int delta)
```
`delta` is overflowing when we use buffers >= 2^31.
This macro is somtimes used to subtract 1 by adding (-1) like here:
https://github.com/open-mpi/ompi/blob/master/ompi/mca/pml/ob1/pml_ob1_recvreq.c#L197

I suggest to create separate macro for that, something like OPAL_THREAD_SUB_SIZE_T and change the signature to:
```C
static inline size_t OPAL_THREAD_ADD_SIZE_T(volatile size_t *addr, size_t delta)
```
With Gilles fix and this one I can run the test with 2^33 buffers.

If nobody minds the basic idea I'd like to do that myself coz I've spend reasonable time pursuing this :).

**Comment by artpol84** [comment]
P.S. sorry, OPAL_THREAD_ADD_SIZE_T is not a macro. I mistyped coz it was written with capitals.

**Comment by artpol84** [comment]
P.P.S
* BTW we had a problem here as well:
https://github.com/open-mpi/ompi/blob/master/ompi/mca/pml/ob1/pml_ob1_sendreq.c#L1134
* I've just checked: 2^34 and 2^35 works too with "No errors" message ;). I won't dare to try 2^36 coz it will take whole memory of my nodes :).

**Comment by ggouaillardet** [comment]
@artpol84 you meant my second patch (minus the printf) right ?
I can do the PR tomorrow, but feel free to do it now if you want.

**Comment by artpol84** [comment]
Yep.

четверг, 3 декабря 2015 г. пользователь Gilles Gouaillardet написал:

> @artpol84 <https://github.com/artpol84> you meant my second patch (minus
> the printf) right ?
> I can do the PR tomorrow, but feel free to do it now if you want.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/open-mpi/ompi/issues/1174#issuecomment-161613855>.
>


-- 
-----
Best regards, Artem Polyakov
(Mobile mail)

**Comment by bosilca** [comment]
https://github.com/open-mpi/ompi/pull/1177

**Comment by bosilca** [comment]
It works for me in a variety of environments (all but OS X). Give it a try and let's move it upstream.

**Comment by jsquyres** [comment]
@bosilca Github pro tip: put "Fixes open-mpi/ompi#1174" in one of the commit messages, and it will auto-close this Github issue when the PR gets merged.
